### PR TITLE
fix fd_frank_quic (bad syscall and FDs being closed)

### PR DESCRIPTION
**Pull request by llamb-jump**
fix solana labs being created in the wrong namespace

### CODE REVIEW

File: src/app/frank/fd_frank_quic.c [hunk 1]
```diff
@@ -133,13 +133,14 @@ static long allow_syscalls[] = {
...
   __NR_getrandom, /* OpenSSL RAND_bytes reads getrandom, temporarily used as part of quic_init to generate a certificate */
+  __NR_sendto,    /* fd_xsk requires sendto */
 };
 
 fd_frank_task_t frank_quic = {
   .name     = "quic",
   .in_wksp  = NULL,
   .out_wksp = "quic_verify",
-  .close_fd_start = 4, /* stdin, stdout, stderr, logfile */
+  .close_fd_start = 6, /* stdin, stdout, stderr, logfile, network namespace, xsk->xsk_fd */
```
Comment (mmcgee-jump) [inline src/app/frank/fd_frank_quic.c:143]: Is the network namespace fd still open here? Or the kernel just doesn't reuse the fd?

File: src/app/fdctl/utility.c [hunk 2]
```diff
@@ -14,19 +14,34 @@
...
-  entered = 1;
+static int namespace_original_fd = 0;
 
+int
+enter_network_namespace( const char * interface ) {
   char path[ PATH_MAX ];
-  snprintf1( path, PATH_MAX, "/var/run/netns/%s", config->tiles.quic.interface );
+  snprintf1( path, PATH_MAX, "/var/run/netns/%s", interface );
+
+  if( !namespace_original_fd ) {
```
Comment (mmcgee-jump) [inline src/app/fdctl/utility.c:24]: if( FD_LIKELY( ... )

File: src/app/fdctl/utility.c [hunk 3]
```diff
@@ -14,19 +14,34 @@
...
+    namespace_original_fd = open( "/proc/self/ns/net", O_RDONLY | O_CLOEXEC );
+    if( FD_UNLIKELY( namespace_original_fd < 0 ) ) FD_LOG_ERR(( "failed to open /proc/self/ns/net (%i-%s)", errno, strerror( errno ) ));
+  }
 
   int fd = open( path, O_RDONLY | O_CLOEXEC );
   if( FD_UNLIKELY( fd < 0 ) ) FD_LOG_ERR(( "failed to open `%s` (%i-%s)", path, errno, strerror( errno ) ));
   if( FD_UNLIKELY( setns( fd, CLONE_NEWNET ) ) ) FD_LOG_ERR(( "failed to enter network namespace `%s` (%i-%s)", path, errno, strerror( errno ) ));
+  close( fd );
```
Comment (mmcgee-jump) [inline src/app/fdctl/utility.c:32]: Error checking, FD_UNLIKELY etc.

File: src/app/fdctl/utility.c [hunk 4]
```diff
@@ -14,19 +14,34 @@
...
+
+  return 0;
```
Comment (mmcgee-jump) [inline src/app/fdctl/utility.c:34]: Doesn't need the return I think

File: src/app/fdctl/utility.c [hunk 5]
```diff
@@ -14,19 +14,34 @@
...
+}
+
+void
+leave_network_namespace() {
+  if( !namespace_original_fd ) return;
```
Comment (mmcgee-jump) [inline src/app/fdctl/utility.c:39]: FD_UNLIKELY

File: src/app/frank/fd_frank_quic.c (hunk 1 above)
Comment (llamb-jump) [reply src/app/frank/fd_frank_quic.c:150]: Updated to close the network namespace FD.

File: src/app/fdctl/utility.c [hunk 6]
```diff
@@ -14,21 +14,37 @@
 #include <linux/netlink.h>
 #include <linux/rtnetlink.h>
 
-static int entered = 0;
+static int namespace_original_fd = 0;
 
 void
-enter_network_namespace( config_t * const config ) {
-  if( FD_LIKELY( !config->development.netns.enabled || entered ) ) return;
```
Comment (mmcgee-jump) [inline src/app/fdctl/utility.c:21]: Why did we remove this check? I think it's an error to enter the namespace if netns is not enabled, that will prevent bind from succeeding on a host device.

File: src/app/frank/fd_frank_quic.c [hunk 7]
```diff
@@ -19,6 +20,12 @@ init( fd_frank_args_t * args ) {
   args->xsk = fd_xsk_join( fd_wksp_pod_map( args->tile_pod, "xsk" ) );
   if( FD_UNLIKELY( !args->xsk ) ) FD_LOG_ERR(( "fd_xsk_join failed" ));
 
+  /* We close utility.c:namespace_original_fd which is opened with enter_network_namespace and
+     closed in leave_network_namespace. This tile does not want to leave the namespace so close the FD
+     instead of leaking it. */
+  int ret = close( 4 );
```
Comment (mmcgee-jump) [inline src/app/frank/fd_frank_quic.c:26]: This feels a bit hacky and modifies the prod boot sequence to deal with development code, which I don't think we want. If you launch the Solana process before any Firedancer tiles, you can do the following sequence.

 1. Leave netns if it's entered (due to running config), this only needs to happen in `dev.c` as fdctl will never do configuration as part of `run`, close the fd.
 2. Spawn Solana tile. Not in network namespace. No fds open.
 3. Rejoin netns from path /var/... if needed.
 4. Close after joining.
 5. Spawn Firedancer tiles, no fds open.

File: src/app/frank/fd_frank_quic.c [hunk 8]
```diff
@@ -133,13 +140,14 @@ static long allow_syscalls[] = {
   __NR_nanosleep, /* fd_tempo_tick_per_ns calibration */
   __NR_getpid,    /* OpenSSL RAND_bytes checks pid, temporarily used as part of quic_init to generate a certificate */
   __NR_getrandom, /* OpenSSL RAND_bytes reads getrandom, temporarily used as part of quic_init to generate a certificate */
+  __NR_sendto,    /* fd_xsk requires sendto */
```
Comment (mmcgee-jump) [inline src/app/frank/fd_frank_quic.c:143]: This should be changed to __NR_send to be maximally restrictive. I'm not sure why that wasn't defined for you, but maybe it's not in syscalls.h and you need another header. The syscall definitely exists.

**Comment by mmcgee-jump** [comment]
Everyone on the team seems to have different standards for the commit messages, but I would maybe make this a bit more verbose, something like

> fdctl: allow send() syscall in quic tile, and fixes for running under network namespace
>
> QUIC uses the `send` syscall internally to cause the device to flush pending reads/writes. This should tend to happen automatically as the driver spins its own loop reading/writing our XSKs, but we have found for now it can improve performance.
>
> As well, when using network namespaces for development, we should not run Solana inside it for now since they don't support this.

File: src/app/fdctl/utility.c (hunk 6 above)
Comment (llamb-jump) [reply src/app/fdctl/utility.c:21]: that was a mistake good catch

File: src/app/frank/fd_frank_quic.c (hunk 8 above)
Comment (llamb-jump) [reply src/app/frank/fd_frank_quic.c:137]: __NR_send isn't available on our current hosts

File: src/app/fdctl/utility.c (hunk 3 above)
Comment (llamb-jump) [reply src/app/fdctl/utility.c]: updated

File: src/app/frank/fd_frank_quic.c (hunk 7 above)
Comment (llamb-jump) [reply src/app/frank/fd_frank_quic.c]: updated

**Review by mmcgee-jump — DISMISSED** [review]

**Review by mmcgee-jump — APPROVED** [review]

**Pull request merged**